polkadot/node/service/src/builder/mod.rs

```diff
@@ -0,0 +1,869 @@
+			// Grandpa performance can be improved a bit by tuning this parameter, see:
+			// https://github.com/paritytech/polkadot/issues/5464
+			gossip_duration: Duration::from_millis(1000),
+			justification_generation_period: GRANDPA_JUSTIFICATION_PERIOD,
+			name: Some(name),
+			observer_enabled: false,
+			keystore: keystore_opt,
+			local_role: role,
+			telemetry: telemetry.as_ref().map(|x| x.handle()),
+			protocol_name: grandpa_protocol_name,
+		};
+
+		let enable_grandpa = !disable_grandpa;
+		if enable_grandpa {
+			// start the full GRANDPA voter
+			// NOTE: unlike in substrate we are currently running the full
+			// GRANDPA voter protocol for all full nodes (regardless of whether
+			// they're validators or not). at this point the full voter should
+			// provide better guarantees of block and vote data availability than
+			// the observer.
+
+			let mut voting_rules_builder = sc_consensus_grandpa::VotingRulesBuilder::default();
+
+			#[cfg(not(feature = "malus"))]
+			let _malus_finality_delay = None;
+
+			if let Some(delay) = _malus_finality_delay {
+				info!(?delay, "Enabling malus finality delay",);
+				voting_rules_builder =
+					voting_rules_builder.add(sc_consensus_grandpa::BeforeBestBlockBy(delay));
+			};
+
+			let grandpa_config = sc_consensus_grandpa::GrandpaParams {
+				config,
+				link: link_half,
+				network: network.clone(),
+				sync: sync_service.clone(),
+				voting_rule: voting_rules_builder.build(),
+				prometheus_registry: prometheus_registry.clone(),
+				shared_voter_state,
+				telemetry: telemetry.as_ref().map(|x| x.handle()),
+				notification_service: grandpa_notification_service,
+				offchain_tx_pool_factory: OffchainTransactionPoolFactory::new(
+					transaction_pool.clone(),
+				),
+			};
+
+			task_manager.spawn_essential_handle().spawn_blocking(
+				"grandpa-voter",
+				None,
+				sc_consensus_grandpa::run_grandpa_voter(grandpa_config)?,
+			);
+		}
+
+		Ok(NewFull {
+			task_manager,
+			client,
+			overseer_handle,
+			network,
+			sync_service,
+			rpc_handlers,
+			backend,
+		})
+	}
+}
+
+/// Create a new full node of arbitrary runtime and executor.
+///
+/// This is an advanced feature and not recommended for general use. Generally, `build_full` is
+/// a better choice.
+///
+/// `workers_path` is used to get the path to the directory where auxiliary worker binaries reside.
+/// If not specified, the main binary's directory is searched first, then `/usr/lib/polkadot` is
+/// searched. If the path points to an executable rather then directory, that executable is used
+/// both as preparation and execution worker (supposed to be used for tests only).
+pub fn new_full<
+	OverseerGenerator: OverseerGen,
+	Network: sc_network::NetworkBackend<Block, <Block as BlockT>::Hash>,
+>(
+	config: Configuration,
+	params: NewFullParams<OverseerGenerator>,
+) -> Result<NewFull, Error> {
+	PolkadotServiceBuilder::<OverseerGenerator, Network>::new(config, params)?.build()
```

Comment:
I like the builder pattern better than `new_full` nicely done here! 🙏 

dq: How would the code look like if we remove the `new_full` fn in favour of strongly typing the builder?

Reply:
That makes sense! My question / suggestion is to deprecate the `new_full` method entirely and rely on the builder.

Ideally, we'd be able to uncouple the `params` from the `config`, but that's not worthwhile at the moment:

```rust

// Instead of `new_full` strongly type the builder:

PolkadotServiceBuilder::new()
   .with_params(params)
   .with_extra_request_resposne(..)
   .build(config);
```

Feel free to ignore this, more of a nit than anything 🙏

Reply:
Agree that this would be cool, but not necessary for now. The usage of this in custom code should also go down with omni-node.